jcl/src/java.base/share/classes/openj9/internal/tools/attach/target/TargetDirectory.java

```diff
@@ -130,33 +135,40 @@ static String createMyDirectory(String myVmId, boolean preserveId) throws IOExce
 	 */
 	static boolean deleteMyFiles() {
 		if (LOGGING_DISABLED != loggingStatus) {
-			IPC.logMessage("deleting my files"); //$NON-NLS-1$
+			IPC.logMessage("deleting my files: attachInfo files and attachNotificationSync files."); //$NON-NLS-1$
 		}
 
 		if ((null != advertisementFileObject) && advertisementFileObject.delete()) {
+			IPC.logMessage("deleted ", advertisementFileObject.getAbsolutePath()); //$NON-NLS-1$
 			advertisementFileObject = null;		
 		} else {
 			return false;
 		}
 		if ((null != syncFileObject) && syncFileObject.delete()) {
+			IPC.logMessage("deleted ", syncFileObject.getAbsolutePath()); //$NON-NLS-1$
 			syncFileObject = null;
 		} else {
 			return false;
 		}
+		if (LOGGING_DISABLED != loggingStatus) {
+			IPC.logMessage("deleted my files with success."); //$NON-NLS-1$
+		}
 		return true;
 	}
 
 	/**
 	 * Remove this VM's target directory and the files within it
 	 * @param directoryEmpty set to true if the directory is supposedly empty
+	 * @vmId target VM ID
 	 */
-	static void deleteMyDirectory(boolean directoryEmpty) {
+	static void deleteMyDirectory(boolean directoryEmpty, String vmId) {
 		if (LOGGING_DISABLED != loggingStatus) {
-			IPC.logMessage("deleting my directory "); //$NON-NLS-1$
+			IPC.logMessage("deleting my directory with given vmId: ", vmId); //$NON-NLS-1$
 		}		
 		/* try the fast path for deleting files and directories. try the heavyweight path if something fails */
 		if ((!directoryEmpty && !deleteMyFiles()) || (null == targetDirectoryFileObject) || !targetDirectoryFileObject.delete()) {
-			deleteTargetDirectory(AttachHandler.getVmId());	
+			IPC.logMessage("fast path for deleting files and dirs fail, trying the heavyweight path for vmId: ", vmId);
```

Comment:
```suggestion
			if (vmId.isEmpty()) {
				vmId = AttachHandler.getVmId();
			}
			IPC.logMessage("fast path for deleting files and dirs fail, trying the heavyweight path for vmId: ", vmId); //$NON-NLS-1$
```

Reply:
Thanks for reply, I am OoO atm, will check today later.
Hope we can close review today :)

Reply:
@JasonFengJ9 
Now, got the point now.
If so, I prefer move such logic to calleee instead inside 
`static void deleteMyDirectory(boolean directoryEmpty, String vmId)`
method, which make it more straightforward and easy for understanding.

On other word, I prefer revert to my previous change：get the vmId by `AttachHandler.getVmId()`  at https://github.com/eclipse-openj9/openj9/blob/ba660c9a9fac1be3dabdadd8ee63a299cb591684/jcl/src/java.base/share/classes/openj9/internal/tools/attach/target/AttachHandler.java#L393

Same for https://github.com/eclipse-openj9/openj9/blob/ba660c9a9fac1be3dabdadd8ee63a299cb591684/jcl/src/java.base/share/classes/openj9/internal/tools/attach/target/AttachHandler.java#L557

What's your idea?

Reply:
The initial discussion https://github.com/eclipse-openj9/openj9/pull/13331#discussion_r692028704 was to introduce an empty `vmId` to delay `AttachHandler.getVmId()` until it is necessary because `AttachHandler.getVmId()` requires a synchronization call.
https://github.com/eclipse-openj9/openj9/blob/cd8db74f1d43b37e3bdbd1c72f9e120d18047634/jcl/src/java.base/share/classes/openj9/internal/tools/attach/target/AttachHandler.java#L823-L825

Alternatively I would suggest following:
```
	/**
	 * A helper to try the fast path for deleting files and directories, and
	 * determine if the heavy-weight path is required.
	 * 
	 * @param directoryEmpty set to true if the directory is supposedly empty
	 * 
	 * @return true if the heavy-weight path is still required, otherwise false.
	 */
	private static boolean isHeavyweightPathDeletionRequired(boolean directoryEmpty) {
		if (LOGGING_DISABLED != loggingStatus) {
			IPC.logMessage("deleting my directory "); //$NON-NLS-1$
		}
		return ((!directoryEmpty && !deleteMyFiles()) || (null == targetDirectoryFileObject)
				|| !targetDirectoryFileObject.delete());
	}

	/**
	 * Remove this VM's target directory and the files within it
	 * 
	 * @param directoryEmpty set to true if the directory is supposedly empty
	 */
	static void deleteMyDirectory(boolean directoryEmpty) {
		if (isHeavyweightPathDeletionRequired(directoryEmpty)) {
			deleteTargetDirectory(AttachHandler.getVmId());
		}
	}

	/**
	 * Remove this VM's target directory and the files within it
	 * 
	 * @param directoryEmpty set to true if the directory is supposedly empty
	 * @vmId target VM ID
	 */
	static void deleteMyDirectory(boolean directoryEmpty, String vmId) {
		if (isHeavyweightPathDeletionRequired(directoryEmpty)) {
			IPC.logMessage("fast path for deleting files and dirs fail, trying the heavyweight path for vmId: ", vmId); //$NON-NLS-1$
			deleteTargetDirectory(vmId);
		}
	}
```

Reply:
What I proposed like below suggested changes.

Reply:
> Hope we can close review today

It might have took longer than expected, but we are getting there, please stay patient :)

Reply:
- The `isHeavyweightPathDeletionRequired(boolean directoryEmpty)` is not realy doing what as its method name, it is trying to delete some path....
-  Not sure you noticed one fact or not.  At `AttachHandler.createFiles(String newDisplayName`  method,   which only called during initialize, the vmId is set at  https://github.com/eclipse-openj9/openj9/blob/cd8db74f1d43b37e3bdbd1c72f9e120d18047634/jcl/src/java.base/share/classes/openj9/internal/tools/attach/target/AttachHandler.java#L214, while, we have to delete the given target dirs and files before that.

How about another proposal which using Functioanl object, and it looks one more simple change, like below 
```
	/**
	 * Remove this VM's target directory and the files within it
	 * @param directoryEmpty set to true if the directory is supposedly empty
	 * @vmId target VM ID
	 */
	static void deleteMyDirectory(boolean directoryEmpty, Supplier<String> vmIdSupplier) {
		/* try the fast path for deleting files and directories. try the heavyweight path if something fails */
		if ((!directoryEmpty && !deleteMyFiles()) || (null == targetDirectoryFileObject) || !targetDirectoryFileObject.delete()) {
			String vmId = vmIdSupplier.get();
			IPC.logMessage("fast path for deleting files and dirs fail, trying the heavyweight path for vmId: ", vmId);
			deleteTargetDirectory(vmId);
		}
	}		
```

Reply:
> we have to delete the given target dirs and files before that.

Changing `preserveId` to `true` seems introduce the possibility to retrieve `vmId` before setting it if the fast path deletion doesn't succeed. I think in that case `pidProperty` passed within `TargetDirectory.createMyDirectory()` can be used, i.e., when `getVmId()` returns an empty string.

Reply:
> > we have to delete the given target dirs and files before that.
> 
> Changing `preserveId` to `true` seems introduce the possibility to retrieve `vmId` before setting it if the fast path deletion doesn't succeed. I think in that case `pidProperty` passed within `TargetDirectory.createMyDirectory()` can be used, i.e., when `getVmId()` returns an empty string.

`vmId` should be setting only initialized with success.
The problem is that the current https://github.com/eclipse-openj9/openj9/blob/ca4973a2b0d9414a726e826c1e5d5506761c395d/jcl/src/java.base/share/classes/openj9/internal/tools/attach/target/TargetDirectory.java#L153
try to delete the target dirs and files via retrieving from `getVmId()` instead of from `pidProperty` regardless which phase(during initialization or not), then finally failure since vmId is not set yet.

My proposed change try to resolve above issue.
Do you have reviewed my latest commit?

Reply:
> try to delete the target dirs and files via retrieving from getVmId() instead of from pidProperty regardless which phase(during initialization or not), then finally failure since vmId is not set yet.

Okay, I see you meant `deleteMyDirectory(false, () -> myVmId)` fixed the problem along with `TargetDirectory.createMyDirectory(pidProperty, true)`.
Again, lambda reference has to be removed.
Using lambda reference during early bootstrap period can change VM behaviours which might be subtle and platform specific, and is not easy to be discovered by sanity functional tests.

Reply:
> The isHeavyweightPathDeletionRequired(boolean directoryEmpty) is not really only doing what as its method name, it is trying to delete some path.... Well, I also like to exact these block to seperate method named deleteMyDirectoryByFastPath

How about `isHeavyweightRequiredAfterFastpathDeletion(boolean directoryEmpty)`?

Reply:
> > try to delete the target dirs and files via retrieving from getVmId() instead of from pidProperty regardless which phase(during initialization or not), then finally failure since vmId is not set yet.
> 
> Okay, I see you meant `deleteMyDirectory(false, () -> myVmId)` fixed the problem along with `TargetDirectory.createMyDirectory(pidProperty, true)`.
> Again, lambda reference has to be removed.
> Using lambda reference during early bootstrap period can change VM behaviours which might be subtle and platform specific, and is not easy to be discovered by sanity functional tests.

Yes, exactly.
As I mentioned below , will remove lambda in next commit for review.

Reply:
> > The isHeavyweightPathDeletionRequired(boolean directoryEmpty) is not really only doing what as its method name, it is trying to delete some path.... Well, I also like to exact these block to seperate method named deleteMyDirectoryByFastPath
> 
> How about `isHeavyweightRequiredAfterFastpathDeletion(boolean directoryEmpty)`?

It's better and OK to me now:)